gen/cycle.ml

```diff
@@ -817,136 +871,163 @@ let set_same_loc st n0 =
+                  (if do_kvm && check_fault then label_fault pte_val else None), st in 
               let bank = n.evt.bank in
-                begin match bank with
-              | Instr -> Warn.fatal "not letting instr write happen"
+              begin match bank with
+              | Instr -> Warn.fatal "instruction annotation to data bank not possible?"
               | Ord ->
-                  let st = CoSt.next_co st bank in
-                  let v = CoSt.get_co st bank in
-                  n.evt <- { n.evt with ins = v;} ;
-                  do_set_write_val next_x_ok st pte_val ns
-               | _ -> do_set_write_val next_x_ok st pte_val ns
-            end
-            end
-        | Some (R|J) |None -> do_set_write_val next_x_ok st pte_val ns
-        end
+                 let st = set_write_val_ord st n in
+                 let check_fault, st = fault_update st in
+                 n.evt <- { n.evt with check_fault; check_value; };
+                 (next_x_ok, st)
+              | Pair ->
+                 (* Same code as for Ord, however notice that
+                    CoSet.set_cell has a case for pairs.
+                    However increment of current value is by 2 *)
+                 let cell = CoSt.get_cell st in
+                 assert (Array.length cell>=2) ;
+                 let st = CoSt.next_co st Ord in (* Pre-increment *)
+                 let st = set_write_val_ord st n in
+                 let check_fault, st = fault_update st in
+                 n.evt <- { n.evt with check_fault; check_value; };
+                 (next_x_ok, st)
+              | Tag|CapaTag|CapaSeal ->
+                 let st = CoSt.next_co st bank in
+                 let v = CoSt.get_co st bank in
+                 n.evt <- { n.evt with v = v; check_value; } ;
+                 let e,st = CoSt.set_tcell st n.evt in
+                 n.evt <- e ;
+                 (next_x_ok, st)
+              | VecReg a ->
+                 let st = CoSt.step_simd st a in
+                 let cell = CoSt.get_cell st in
+                 let vecreg  = E.SIMD.read a cell in
+                 let v =
+                   match vecreg with
+                     | (v::_)::_ -> v
+                     | _ -> assert false in
+                 n.evt <- { n.evt with vecreg; cell; v; check_value; } ;
+                 (next_x_ok, st)
+              | Pte ->
+              (* TODO Rework here, esp the function `next_loc` and ref value `next_x_pred`.
+                 They are all difficult to understand. *)
+                 let next_x_pred = ref false in
+                 (* get the previous `pte_value` *)
+                 let pte_val = CoSt.get_pte_value st in
+                 (* update the pte value in kvm variant *)
+                 let pte_val =
+                   if do_kvm then begin
+                       let next_loc () =
+                         match n.evt.loc with
+                         | Code.Data x ->
+                            begin try
```

Comment:
Would it be worth putting this in a seperate function to improve readability in this function?

Reply:
I did not change this part, though I will have a look what I can do.

Reply:
I will do this as a separate merge, once I find a solution to make it more readable.